src/Model/User.php

```diff
@@ -304,100 +280,83 @@ public static function create(array $data)
-			dba::delete('user', array('uid' => $newuid));
-			return $result;
+		if (!$uid) {
+			throw new Exception(t('An error occurred during registration. Please try again.'));
 		}
 
-		if (x($newuid) !== false) {
-			$r = q("INSERT INTO `profile` ( `uid`, `profile-name`, `is-default`, `name`, `photo`, `thumb`, `publish`, `net-publish` )
-				VALUES ( %d, '%s', %d, '%s', '%s', '%s', %d, %d ) ",
-				intval($newuid),
-				t('default'),
-				1,
-				dbesc($username),
-				dbesc(System::baseUrl() . "/photo/profile/{$newuid}.jpg"),
-				dbesc(System::baseUrl() . "/photo/avatar/{$newuid}.jpg"),
-				intval($publish),
-				intval($netpublish)
-			);
-			if ($r === false) {
-				$result['message'] .= t('An error occurred creating your default profile. Please try again.') . EOL;
-				// Start fresh next time.
-				dba::delete('user', array('uid' => $newuid));
-				return $result;
-			}
+		// if somebody clicked submit twice very quickly, they could end up with two accounts
+		// due to race condition. Remove this one.
+		$user_count = dba::count('user', ['nickname' => $nickname]);
+		if ($user_count > 1) {
+			dba::delete('user', ['uid' => $uid]);
 
-			// Create the self contact
-			Contact::createSelfFromUserId($newuid);
+			throw new Exception(t('Nickname is already registered. Please choose another.'));
+		}
 
-			// Create a group with no members. This allows somebody to use it
-			// right away as a default group for new contacts.
-			Group::create($newuid, t('Friends'));
+		$insert_result = dba::insert('profile', [
+			'uid' => $uid,
+			'name' => $username,
+			'photo' => System::baseUrl() . "/photo/profile/{$uid}.jpg",
+			'thumb' => System::baseUrl() . "/photo/avatar/{$uid}.jpg",
+			'publish' => $publish,
+			'is-default' => 1,
+			'net-publish' => $netpublish,
+			'profile-name' => t('default')
+		]);
+		if (!$insert_result) {
+			dba::delete('user', ['uid' => $uid]);
+
+			throw new Exception(t('An error occurred creating your default profile. Please try again.'));
+		}
 
-			$r = q("SELECT `id` FROM `group` WHERE `uid` = %d AND `name` = '%s'",
-				intval($newuid),
-				dbesc(t('Friends'))
-			);
-			if (DBM::is_result($r)) {
-				$def_gid = $r[0]['id'];
+		// Create the self contact
+		if (!Contact::createSelfFromUserId($uid)) {
+			dba::delete('user', ['uid' => $uid]);
 
-				q("UPDATE `user` SET `def_gid` = %d WHERE `uid` = %d",
-					intval($r[0]['id']),
-					intval($newuid)
-				);
-			}
+			throw new Exception(t('An error occurred creating your self contact. Please try again.'));
+		}
 
-			if (Config::get('system', 'newuser_private') && $def_gid) {
-				q("UPDATE `user` SET `allow_gid` = '%s' WHERE `uid` = %d",
-					dbesc("<" . $def_gid . ">"),
-					intval($newuid)
-				);
-			}
+		// Create a group with no members. This allows somebody to use it
+		// right away as a default group for new contacts.
+		$def_gid = Group::create($uid, t('Friends'));
+		if (!$def_gid) {
+			dba::delete('user', ['uid' => $uid]);
```

Comment:
Packing this into a transaction and using "commit" if everything was okay - or "rollback" if not - would be the cleanest way. (But of course this here is better than what had been done before)

Reply:
You're again absolutely right.